src/Tasks/GetAssembliesMetadata.cs

```diff
@@ -0,0 +1,120 @@
+        {
+            get => _assemblyPaths;
+
+            set
+            {
+                ErrorUtilities.VerifyThrowArgumentNull(value, nameof(AssemblyPaths));
+                _assemblyPaths = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets a list of resolved assembly metadata.
+        /// </summary>
+        [Output]
+        public ITaskItem[] AssembliesMetadata => _assembliesMetadata;
+
+        /// <summary>
+        /// Sets metadata on the assembly path.
+        /// </summary>
+        private TaskItem CreateItemWithMetadata(AssemblyAttributes attributes)
+        {
+            TaskItem referenceItem = new()
+            {
+                ItemSpec = attributes.AssemblyFullPath,
+            };
+
+            IMetadataContainer referenceItemAsMetadataContainer = referenceItem;
+            referenceItemAsMetadataContainer.ImportMetadata(EnumerateCommonMetadata());
+
+            return referenceItem;
+
+            IEnumerable<KeyValuePair<string, string>> EnumerateCommonMetadata()
+            {
+                yield return new KeyValuePair<string, string>(nameof(attributes.AssemblyName), attributes.AssemblyName);
+                yield return new KeyValuePair<string, string>(nameof(attributes.RuntimeVersion), attributes.RuntimeVersion);
+                yield return new KeyValuePair<string, string>(nameof(attributes.RevisionNumber), attributes.RevisionNumber.ToString());
+                yield return new KeyValuePair<string, string>(nameof(attributes.BuildNumber), attributes.BuildNumber.ToString());
+                yield return new KeyValuePair<string, string>(nameof(attributes.Culture), attributes.Culture);
+                yield return new KeyValuePair<string, string>(nameof(attributes.DefaultAlias), attributes.DefaultAlias);
+                yield return new KeyValuePair<string, string>(nameof(attributes.Description), attributes.Description);
+                yield return new KeyValuePair<string, string>(nameof(attributes.Guid), attributes.Guid);
+                yield return new KeyValuePair<string, string>(nameof(attributes.MajorVersion), attributes.MajorVersion.ToString());
+                yield return new KeyValuePair<string, string>(nameof(attributes.MinorVersion), attributes.MinorVersion.ToString());
+                yield return new KeyValuePair<string, string>(nameof(attributes.PeKind), attributes.PeKind.ToString());
+                yield return new KeyValuePair<string, string>(nameof(attributes.PublicKey), attributes.PublicKey);
+                yield return new KeyValuePair<string, string>(nameof(attributes.IsAssembly), attributes.IsAssembly.ToString());
+                yield return new KeyValuePair<string, string>(nameof(attributes.TargetFrameworkMoniker), attributes.TargetFrameworkMoniker);
+                yield return new KeyValuePair<string, string>(nameof(attributes.IsImportedFromTypeLib), attributes.IsImportedFromTypeLib.ToString());
+            }
```

Comment:
```suggestion
            Dictionary<string, string> metadata = new()
            {
                { nameof(attributes.AssemblyName), attributes.AssemblyName },
                { nameof(attributes.RuntimeVersion), attributes.RuntimeVersion },
                { nameof(attributes.RevisionNumber), attributes.RevisionNumber.ToString() },
                { nameof(attributes.BuildNumber), attributes.BuildNumber.ToString() },
                { nameof(attributes.Culture), attributes.Culture },
                { nameof(attributes.DefaultAlias), attributes.DefaultAlias },
                { nameof(attributes.Description), attributes.Description },
                { nameof(attributes.Guid), attributes.Guid },
                { nameof(attributes.MajorVersion), attributes.MajorVersion.ToString() },
                { nameof(attributes.MinorVersion), attributes.MinorVersion.ToString() },
                { nameof(attributes.PeKind), attributes.PeKind.ToString() },
                { nameof(attributes.PublicKey), attributes.PublicKey },
                { nameof(attributes.IsAssembly), attributes.IsAssembly.ToString() },
                { nameof(attributes.TargetFrameworkMoniker), attributes.TargetFrameworkMoniker },
                { nameof(attributes.IsImportedFromTypeLib), attributes.IsImportedFromTypeLib.ToString() },
            };
```

Reply:
yield usage was inspired by this part of code
https://github.com/dotnet/msbuild/blob/a9341111228592c25c43ecf6858e7b39e525b5a1/src/Tasks/AssemblyDependency/ReferenceTable.cs#L2860

Reply:
Jan's suggestion looks good - although it would be interesting to measure. My guess is that the `yield` version may allocate less but the `Dictionary` one would be faster.